### PR TITLE
Correct Java type of conditional expressions in dataflow.

### PR DESCRIPTION
**Pull request by smillst**

### CODE REVIEW

**Review by mernst — APPROVED** [review]
Thanks!

**Comment by smillst** [comment]
@mernst I had to re-do the implementation because I had not accounted for CFValue subclasses.  Could you re-review?

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java [hunk 1]
```diff
@@ -480,16 +507,27 @@ private V upperBound(@Nullable V other, boolean shouldWiden) {
...
+    return upperBound(other, lubTypeMirror, shouldWiden);
+  }
 
+  /**
+   * Returns an upper bound of this and {@code other}.
+   *
+   * @param other an abstract value
+   * @param upperBoundTypeMirror the underlying type of the returned value
+   * @param shouldWiden true if the lub should perform widening
+   * @return the least upper bound of this and {@code other}
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java:519]: Above it says "an upper bound", here it says "the least upper bound".  Please clarify.
Also, please write a comment discussing the relationship between this method and `leastUpperBound`.

**Review by mernst — CHANGES_REQUESTED** [review]

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java [hunk 2]
```diff
@@ -432,11 +432,34 @@ private final AnnotationMirror lowestQualifier(AnnotationMirror qual1, Annotatio
...
   @Override
-  public V leastUpperBound(@Nullable V other) {
+  public final V leastUpperBound(@Nullable V other) {
     return upperBound(other, false);
   }
 
+  /**
+   * Compute the least upper bound of two values. The returned value with have a Java type of {@code
+   * typeMirror}. {@code TypeMirror} should be an upper bound of the Java types of {@code this} an
+   * {@code other}, but it does not have be to the leaster upper bound.
```
Comment (msridhar) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java:449]: Small typo here:
```suggestion
   * {@code other}, but it does not have be to the least upper bound.
```

Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java]: Thanks!

File: checker/src/main/java/org/checkerframework/checker/nullness/KeyForValue.java [hunk 3]
```diff
@@ -84,8 +84,10 @@ public KeyForValue(
   }
 
   @Override
-  public KeyForValue leastUpperBound(@Nullable KeyForValue other) {
-    KeyForValue lub = super.leastUpperBound(other);
+  protected KeyForValue upperBound(
+      @Nullable KeyForValue other, TypeMirror upperBoundTypeMirror, boolean shouldWiden) {
+    KeyForValue lub = super.upperBound(other, upperBoundTypeMirror, shouldWiden);
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/nullness/KeyForValue.java:89]: This variable name suggests that `upperBound()` actually returns the least upper bound.  Why is `least` not in its name?

OK, I see now.  I wonder if its name should be `leastUpperBoundOrWiden()`, for clarity.

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 4]
```diff
@@ -668,8 +668,13 @@ public TransferResult<V, S> visitTernaryExpression(
     V elseValue = p.getValueOfSubNode(n.getElseOperand());
     V resultValue = null;
     if (thenValue != null && elseValue != null) {
+      // If a conditional expression is a poly expression, then its Java type is the type of its
+      // context. (For example, the type of the condition expression is in `Object o = b ? "" : "";`
+      // is `Object`, not `String`.)
+      // So, use the Java type of the conditional expression and the annotations for each branch.
+      TypeMirror ternaryType = TreeUtils.typeOf(n.getTree());
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractTransfer.java:675]: It's a bit inconsistent to use "contitional" in the comment but "ternary" in the variable name.

File: framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java [hunk 5]
```diff
@@ -480,16 +506,32 @@ private V upperBound(@Nullable V other, boolean shouldWiden) {
...
+   * Returns an upper bound of {@code this} and {@code other}. The underlying type of the value
+   * returned, is {@code upperBoundTypeMirror}. If {@code shouldWiden} is false, this method returns
+   * the least upper bound of {@code this} and {@code other}.
+   *
+   * <p>This is the implementation of {@link #leastUpperBound(CFAbstractValue, TypeMirror)}, {@link
+   * #leastUpperBound(CFAbstractValue)}, and {@link #upperBound(CFAbstractValue, boolean)}.
+   *
+   * @param other an abstract value
+   * @param upperBoundTypeMirror the underlying type of the returned value
+   * @param shouldWiden true if the lub should perform widening
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/flow/CFAbstractValue.java:522]: It's not really a lub.

File: framework/src/main/java/org/checkerframework/framework/util/AtmLubVisitor.java [hunk 6]
```diff
@@ -91,7 +91,7 @@ private AnnotatedTypeMirror lubWithNull(
     } else {
       otherAsLub = AnnotatedTypes.asSuper(atypeFactory, otherType, lub);
     }
-    TypeMirror otherTM = otherAsLub.getUnderlyingType();
+    TypeMirror otherTM = otherType.getUnderlyingType();
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/util/AtmLubVisitor.java:94]: Is this logically part of this pull request?  Should it be committed in a separate PR?

**Review by mernst — CHANGES_REQUESTED** [review]

**Review by mernst — APPROVED** [review]

**Pull request merged**